core/lib/zksync_core/src/metadata_calculator/recovery.rs

```diff
@@ -0,0 +1,617 @@
+        let health = Health::from(HealthStatus::Ready).with_details(RecoveryMerkleTreeInfo {
+            mode: "recovery",
+            chunk_count: self.chunk_count,
+            recovered_chunk_count,
+        });
+        self.inner.update(health);
+    }
+}
+
+#[derive(Debug)]
+struct RecoveryOptions<'a> {
+    chunk_count: usize,
+    events: Box<dyn HandleRecoveryEvent + 'a>,
+}
+
+impl GenericAsyncTree {
+    /// Ensures that the tree is ready for the normal operation, recovering it from a Postgres snapshot
+    /// if necessary.
+    pub async fn ensure_ready(
+        self,
+        pool: &ConnectionPool,
+        stop_receiver: &watch::Receiver<bool>,
+        health_updater: &HealthUpdater,
+    ) -> anyhow::Result<Option<AsyncTree>> {
+        let (tree, l1_batch) = match self {
+            Self::Ready(tree) => return Ok(Some(tree)),
+            Self::Recovering(tree) => {
+                let l1_batch = snapshot_l1_batch(pool).await?.context(
+                    "Merkle tree is recovering, but Postgres doesn't contain snapshot L1 batch",
+                )?;
+                let recovered_version = tree.recovered_version();
+                anyhow::ensure!(
+                    u64::from(l1_batch.0) == recovered_version,
+                    "Snapshot L1 batch in Postgres ({l1_batch}) differs from the recovered Merkle tree version \
+                     ({recovered_version})"
+                );
+                tracing::info!("Resuming tree recovery with snapshot L1 batch #{l1_batch}");
+                (tree, l1_batch)
+            }
+            Self::Empty { db, mode } => {
+                if let Some(l1_batch) = snapshot_l1_batch(pool).await? {
+                    tracing::info!(
+                        "Starting Merkle tree recovery with snapshot L1 batch #{l1_batch}"
+                    );
+                    let tree = AsyncTreeRecovery::new(db, l1_batch.0.into(), mode);
+                    (tree, l1_batch)
+                } else {
+                    // Start the tree from scratch. The genesis block will be filled in `TreeUpdater::loop_updating_tree()`.
+                    return Ok(Some(AsyncTree::new(db, mode)));
+                }
+            }
+        };
+
+        // FIXME: make choice of ranges more intelligent. NB: ranges must be the same for the entire recovery!
+        let recovery_options = RecoveryOptions {
+            chunk_count: 256,
+            events: Box::new(RecoveryHealthUpdater::new(health_updater)),
+        };
+        tree.recover(l1_batch, recovery_options, pool, stop_receiver)
+            .await
+    }
+}
+
+impl AsyncTreeRecovery {
+    async fn recover(
+        mut self,
+        l1_batch: L1BatchNumber,
+        mut options: RecoveryOptions<'_>,
+        pool: &ConnectionPool,
+        stop_receiver: &watch::Receiver<bool>,
+    ) -> anyhow::Result<Option<AsyncTree>> {
+        let mut storage = pool.access_storage().await?;
+        let (_, snapshot_miniblock) = storage
```

Comment:
Not entirely sure whether `snapshot_miniblock` can be obtained in this way, but the "real" way of getting it shouldn't be much harder.